### DISCUSSION TITLE
Question: How do you adjust the value of a string fact outside Grule?

### DISCUSSION THREAD

**Discussion by davedbase** [hyperjumptech/grule-rule-engine#188] [Q&A]
I'd like to change the incoming value of a fact (a string value). This is what I'm doing:
```go
val, err := exp.GetFact("StringVarName").GetValue()
if err == nil {
	val.SetString("This is my new value")
}
```
I'm not quite sure why it keeps panicking with:
```bash
panic: reflect: reflect.Value.SetString using unaddressable value
```
This is likely a pretty amateur issue, apologies. I'm not very strong with Go reflection yet.

Upon setting the value I'd like to Execute() a reval and see how it impacts my KB.

**Comment by newm4n** [comment]
If your value (reflect.Value) is a value that you get from something like `reflect.ValueOf("some string")` then the value is referencing to a string constant, as string is immutable then you know that you can't change immutable value (not addressable).

The same if you have a variable contains a string, like ` str := "some string"`  and you get the value `reflect.ValueOf(str)` the same rule apply as the variable is referencing to an immutable string. You can make it addressable like `reflect.ValueOf(&str)`, then you obtain a value referensing to the address of that string. After that, you can change the address to point to another string you want the address to point to.

**Reply by davedbase** [reply]
The value is being set with `.Add("MyVar", "TheValueHere")` so does that mean that Grule is adding it as a *string? If so if I'm understanding this correctly I would create a new reflect.Value as *string and change the address of the val to the new location? Apologies if I'm not getting this right away, this is like meta-reflection and I'm having a hard time tracing the logic rn.

**Reply by HymnsForDisco** [reply]
Coming from this Stack Overflow question: https://stackoverflow.com/questions/67236727/reflect-value-string-not-changing-with-setstring

Just wanted to add that the immutability of `string` is actually a red herring here.  Since `reflect.ValueOf` takes a parameter of type `interface{}`, no matter what value is passed to it, that **direct value** gets copied and the copy becomes "boxed" into an interface, where it resides in non-addressable memory (no matter what data type it is).  The reason why I say "direct value" is because values pointed to by the direct value are still held in normal addressable memory.

**Reply by davedbase** [reply]
Thanks for dropping in @HymnsForDisco. Truly appreciate it. If this weren't so rooted in the library I might have gotten through it. Just struggling to wrap my head around it, so I appreciate the patience.

So I attempted using .Add() instead of re-pointing the value of the reflection and that works. I suppose @newm4n my question becomes should I be concerned about any negative symptoms or issues related to solving it this way? I would have done it and not bothered to ask the question but I'm concerned if Grule would have any symptoms that I should be aware of? Any gotchas here?
